src/lib/operators/aggregate_hash.hpp

```diff
@@ -59,62 +60,26 @@ Optionally, the result may also contain:
 - a set of DISTINCT values OR
 - secondary aggregates, which are currently only used by STDDEV_SAMP
 */
-template <typename ColumnDataType, typename AggregateType>
+template <typename ColumnDataType, AggregateFunction aggregate_function>
 class AggregateResult {
  public:
-  std::optional<AggregateType> current_primary_aggregate;
-  size_t aggregate_count = 0;
-  RowID row_id;
+  using AggregateType = typename AggregateTraits<ColumnDataType, aggregate_function>::AggregateType;
 
   using DistinctValues = tsl::robin_set<ColumnDataType, std::hash<ColumnDataType>, std::equal_to<ColumnDataType>,
                                         PolymorphicAllocator<ColumnDataType>>;
 
-  // SecondaryAggregates and DistinctValues are unused in most cases. We store them in a separate variant that is
-  // initialized as needed. This saves us a lot of memory in this critical data structure. Before using this variant
-  // the corresponding ensure_*_initialized method has to be called.
-  using Details = std::variant<SecondaryAggregates<AggregateType>, DistinctValues>;
-
-  void ensure_distinct_values_initialized(boost::container::pmr::monotonic_buffer_resource& buffer) {
-    if (_details) return;
-    _details = std::make_unique<Details>(DistinctValues{&buffer});
-  }
-
-  void ensure_secondary_aggregates_initialized(boost::container::pmr::monotonic_buffer_resource& buffer) {
-    if (_details) return;
-
-    // For some reason, we can't pass &buffer into SecondaryAggregates. Boost does something weird with the allocator.
-    // As SecondaryAggregates does not allocate memory right now anyway (but might once we support additional aggregate
-    // functions), we just use the default allocator for now.
-    _details = std::make_unique<Details>(SecondaryAggregates<AggregateType>{});
-  }
-
-  DistinctValues& distinct_values() {
-    DebugAssert(_details, "AggregateResult::_details has not been properly initialized for this aggregation function");
-    return std::get<DistinctValues>(*_details);
-  }
-
-  const DistinctValues& distinct_values() const {
-    DebugAssert(_details, "AggregateResult::_details has not been properly initialized for this aggregation function");
-    return std::get<DistinctValues>(*_details);
-  }
-
-  SecondaryAggregates<AggregateType>& current_secondary_aggregates() {
-    DebugAssert(_details, "AggregateResult::_details has not been properly initialized for this aggregation function");
-    return std::get<SecondaryAggregates<AggregateType>>(*_details);
-  }
-
-  const SecondaryAggregates<AggregateType>& current_secondary_aggregates() const {
-    DebugAssert(_details, "AggregateResult::_details has not been properly initialized for this aggregation function");
-    return std::get<SecondaryAggregates<AggregateType>>(*_details);
-  }
+  using AccumulatorType = std::conditional_t<
```

Comment:
A rough description and the intent of this data structure would help.

Reply:
What do you mean? There are three lines of comment just above it. I could interleave it with the code to make the structure clearer, but I am not sure what information you are missing.

Reply:
The lines above the `AggregateResult`, but not the `Accumulator`. I just thought that your explanation in the PR description would help here: `specialize per agg function (mostly just simple counter), sometimes more, to save some bytes (24) as that has shown to be quite important for performance`.

Reply:
I'll copy it here, but the comment clearly references the accumulator:

<img width="794" alt="Screenshot 2020-11-06 at 15 17 26" src="https://user-images.githubusercontent.com/575106/98376028-39d39f80-2043-11eb-8951-bbaf73210cfb.png">

Reply:
Ahhh ... now I see it. Don't know what's going on here...

Reply:
Decided against the commnt with the bytes. We should not document what has been done wrongly in the past. Choosing (only) the fitting data type makes sense even before looking at the size of the thing.

Reply:
Now:

```
  // Find the correct accumulator type using conditional_t.
  using AccumulatorType = std::conditional_t<
      // For StandardDeviationSample, use StandardDeviationSampleData as the accumulator,
      aggregate_function == AggregateFunction::StandardDeviationSample, StandardDeviationSampleData,
      // for CountDistinct, use DistinctValues, otherwise use AggregateType
      std::conditional_t<aggregate_function == AggregateFunction::CountDistinct, DistinctValues, AggregateType>>;
```

Reply:
I still think the byte info would help as I would have never cared about a std::optional in this struct. How can such an optional hurt anybody? But that's your call.

Reply:
added 

> Note that the size of this struct is a significant performance factor (see #2252). Be careful when adding fields or changing data types.